theano/gof/tests/test_vm.py

```diff
@@ -371,3 +374,48 @@ def check_storage(storage_map):
         assert check_storage(storage_map)[0]
         assert len(set([id(v) for v in
                         storage_map.values()])) < len(storage_map)
+
+
+def test_shape_reallocation():
+        a, b, c, d = [tensor.dvector(n) for n in ['a', 'b', 'c', 'd']]
+        z = 3*a + b
+        x = c + 4*d
+        y = z + x
+
+        m = theano.compile.get_mode(theano.Mode(linker='vm_nogc'))
+        m_1 = m.excluding('fusion', 'inplace')
+        m_2 = m_1.excluding('ShapeOpt')
+
+        m_dict = {m_1: True, m_2: False}
+
+        for m in m_dict.keys():
+            f = theano.function([a, b, c, d], y, name="test_reduce_memory",
+                                mode=m)
+
+            output = f([1, 2], [3, 5], [9, 8], [4, 3])
+            assert output.any()
+            storage_map = f.fn.storage_map
+
+            def check_storage(storage_map):
+                for i in storage_map.keys():
+                    if not isinstance(i, TensorConstant):
+                        keys_copy = storage_map.keys()[:]
+                        keys_copy.remove(i)
+                        for o in keys_copy:
+                            if isinstance(
```

Comment:
You are right, that is why I made the check again None before calling this. If it isn't None, in this particular test, they must be ndarray, so it would be fine. But we can't reuse that code if the graph start to contain arbitrary object. This make the test code much clearning, it took me a few minutes to understand what was going on.